edb/server/protocol/ai_ext.py

```diff
@@ -1591,19 +1949,27 @@ async def _handle_rag_request(
             if custom_prompt:
                 if not isinstance(custom_prompt, list):
                     raise TypeError(
-                        "prompt.custom must be a list of {role, content} "
-                        "objects"
+                        """prompt.custom must be a list, where each element is
+                            one of the following types:\n
+                        - system message: { role: 'system', content: str }\n
+                        - user message: { role: 'user', content:
+                            [{ type: 'text', text: str }] }\n
+                        - assistant message: { role: 'assistant', content: str,
+                            optional tool_calls: [{id: str, type: 'function',
+                            function: { name: str, arguments: str }}] }\n
+                        - tool message:
+                            { role: 'tool', content: str, tool_call_id: str }"""
                     )
                 for entry in custom_prompt:
                     if (
                         not isinstance(entry, dict)
                         or not entry.get("role")
-                        or not entry.get("content")
-                        or len(entry) > 2
+                        # content can be empty string too
+                        or "content" not in entry
+                        or len(entry) > 3
```

Comment:
why is it ok to have `len(entry) == 3`?

Reply:
one of the following types:\n
                            - system message: { role: 'system', content: str }\n
                            - user message: { role: 'user', content:
                                [{ type: 'text', text: str }] }\n
                            - assistant message: { role: 'assistant', content: str,
                                optional tool_calls: [{id: str, type: 'function',
                                function: { name: str, arguments: str }}] }\n
                            - tool message:
                                { role: 'tool', content: str, tool_call_id: str }"""
                        )
```
the thing is that the message can take few shapes, role and content should always be there but there can be `tool_call_id` or `tool_calls` too depending on the message role. Maybe I should add all these somehow to the if part?

Reply:
I think that might make the intent clearer.

Perhaps worth just checking all the data:
```py
if not isinstance(entry, dict) or not entry.get('role'):
    raise Something

entry_role = entry.get('role')
if entry_role == 'system':
    if set(entry.keys()) != {'role', content'}:
        raise Something
elif entry_role == 'user':
# ...
```